llbcache/resolveimage.go

```diff
@@ -0,0 +1,114 @@
+package llbcache
+
+import (
+	"context"
+	"sync"
+
+	"github.com/moby/buildkit/client/llb"
+	"github.com/opencontainers/go-digest"
+)
+
+// ResolveImageHandle provides an interface for handles to cached
+type ResolveImageHandle interface {
+	// Fetch returns the image digest, image config and possible error that
+	// occurred while looking up the image.  Fetch should be called on
+	// handles returned from Lookup when the `found` bool is true.
+	Fetch(ctx context.Context) (digest.Digest, []byte, error)
+
+	// Store will update the cache handle with the argument values.  Store
+	// should be called on handles return where the `found` bool was true
+	// returned from Lookup.
+	Store(digest.Digest, []byte, error)
+}
+
+// ResolveImageCache provides an interface to lookup resolved images in a cache
+// which allows multiple (and/or parallel) lookups for the same image to be
+// dedup'd and cached.
+type ResolveImageCache interface {
+	// Lookup will look in the cache to find a previously attempted resolve and
+	// will return a handle to that resolve, along with `true` for the `found`
+	// bool.  If the resolve is in-flight then Fetch on the handle will block
+	// until the resolve has completed.  If `false` is returned for the `found`
+	// bool, then the handle Store must be called after performing the resolve.
+	Lookup(ref string, opt llb.ResolveImageConfigOpt) (h ResolveImageHandle, found bool)
+}
+
+type resolveImageCacheKey struct {
+	resolveType llb.ResolverType
+	ref         string
+	os          string
+	arch        string
+	variant     string
+	mode        string
+	store       llb.ResolveImageConfigOptStore
+}
+
+type resolveImageCacheValue struct {
+	digest   digest.Digest
+	config   []byte
+	err      error
+	inflight chan struct{}
+}
+
+// NewResolveImageCache will return a new cache for resolved images.
+func NewResolveImageCache() ResolveImageCache {
+	return &resolveImageCache{
+		cache: map[resolveImageCacheKey]*resolveImageCacheValue{},
+	}
+}
+
+type resolveImageCache struct {
+	mu    sync.RWMutex
+	cache map[resolveImageCacheKey]*resolveImageCacheValue
```

Comment:
I don't think there's any advantage to making this a `RWMutex`. It's only held for an extremely short amount of time and a read lock holder may have to re-lock it for writes.

Reply:
yeah, refactored a bit, is cleaner now.